pilot/pkg/proxy/envoy/v2/grpc.go

```diff
@@ -0,0 +1,262 @@
+					return err
+				}
+			}
+		}
+		ll := &xdsapi.Listener{
+			Name: name,
+		}
+		// TODO: optimize, use a map in pre-computation of sidecar scope
+		for _, el := range con.node.SidecarScope.EgressListeners {
+			for _, s := range el.Services() {
+				if s.Hostname.Matches(host.Name(hostn)) {
+					ll.Address = &envoycore.Address{
+						Address: &envoycore.Address_SocketAddress{
+							SocketAddress: &envoycore.SocketAddress{
+								Address: s.Address,
+								PortSpecifier: &envoycore.SocketAddress_PortValue{
+									PortValue: uint32(portI),
+								},
+							},
+						},
+					}
+					ll.ApiListener = &envoy_config_listener_v2.ApiListener{
+						ApiListener: util.MessageToAny(&v2.HttpConnectionManager{
+							RouteSpecifier: &v2.HttpConnectionManager_Rds{
+								Rds: &v2.Rds {
+									RouteConfigName: name,
+								},
+							},
+						}),
+					}
+				}
+			}
+		}
+
+		lr := util.MessageToAny(ll)
+		resp.Resources = append(resp.Resources, lr)
+	}
+
+	err = con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return err
+	}
+
+	return nil
+}
+
+// Handle a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources.
+func (s *DiscoveryServer) handleAPICDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	if s.handleAck(con, req) {
+		return true
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     ClusterType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+
+	// gRPC doesn't currently support any of the APIs - returning just the expected EDS result.
+	// Since the code is relatively strict - we'll add info as needed.
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		rc := &xdsapi.Cluster{
+			Name: n,
+			ClusterDiscoveryType: &xdsapi.Cluster_Type{Type: xdsapi.Cluster_EDS},
+			EdsClusterConfig: &xdsapi.Cluster_EdsClusterConfig{
+				ServiceName: "outbound|" + portn + "||" + hn,
+				EdsConfig: &envoycore.ConfigSource{
+					ConfigSourceSpecifier: &envoycore.ConfigSource_Ads{
+						Ads: &envoycore.AggregatedConfigSource{},
+					},
+
+				},
+			},
+		}
+		rr := util.MessageToAny(rc)
+		resp.Resources = append(resp.Resources, rr)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return true
+	}
+	return true
+}
+
+// handleSplitRDS supports per-VIP routes, as used by GRPC.
+// This mode is indicated by using names containing full host:port instead of just port.
+// Returns true of the request is of this type.
+func (s *DiscoveryServer) handleSplitRDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	for _, n := range req.ResourceNames {
+		if !strings.Contains(n, ":") {
```

Comment:
This seems like a pretty fragile way to determine this. My first thought would be to use the user agent of gRPC to determine what type we handle. Are there any issues with this?

Reply:
As I mentioned elsewhere, I don't think it's a good idea to make the xDS server's behavior depend on the client's user agent.  Nothing in the protocol should depend on that.

I'm not familiar enough with the Istio code to say whether the approach used here is a good one or not.  I'm just pointing out that the user-agent approach doesn't seem right as an alternative.

Reply:
Unless the client implements the full XDS spec, I don't see how we can be generic otherwise. For example, from the spec `The lb_policy field must be set to ROUND_ROBIN.`. So we need some way to distinguish this from a fully implemented Envoy.

Reply:
Yes, that's part of the WIP - I will add a node metadata to indicate client is in a 'xds-api' mode.

Reply:
I actually don't think an xDS server can automatically assume that all clients support everything in the protocol.  Even Envoy can support different sets of filters based on how it's built.

For gRPC, we're currently in a sort of transition phase, because we're just starting to add xDS features, and it will be a while before they're all there.  But in general, for things that we expect gRPC will *never* support, we should reflect those via [client capabilities](https://www.envoyproxy.io/docs/envoy/latest/api/client_features).

Reply:
Yes but we have strict bounds on the Envoys we support. And we make certain assumptions throughout our code with regard to the envoy versions we support, not to mention we ship the envoys as well. So, I agree with @howardjohn . Shoehorning an entirely different class of XDS receiver into some battle tested code is neither clean nor maintainable in the long run for us. We are happy to support other XDS implementations but the code needs to be both isolated for now until we gain confidence in how the system behaves.